schism/keybinds.c

```diff
@@ -0,0 +1,499 @@
+
+        int char_result = parse_shortcut_character(bind, trimmed, &character);
+        if (char_result == 1) break;
+        if (char_result == -1) { has_problem = 1; break; }
+    }
+
+    if (trimmed) free(trimmed);
+    free(shortcut_dup);
+    if(has_problem) return;
+    keybinds_add_bind_shortcut(bind, key_code, scan_code, character, mods);
+}
+
+void keybinds_parse_shortcut(keybind_bind* bind, const char* shortcut)
+{
+    char* shortcut_dup = strdup(shortcut);
+    char *strtok_ptr;
+    const char* delim = ",";
+
+    for(int i = 0; i < 10; i++) {
+        const char* next = strtok_r(i == 0 ? shortcut_dup : NULL, delim, &strtok_ptr);
+        if (next == NULL) break;
+        keybinds_parse_shortcut_splitted(bind, next);
+    }
+
+    free(shortcut_dup);
+}
+
+/* --------------------------------------------------------------------- */
+/* Initiating keybinds */
+
+void keybinds_add_bind_shortcut(keybind_bind* bind, SDL_Keycode keycode, SDL_Scancode scancode, const char* character, SDL_Keymod modifier)
+{
+    if(bind->shortcuts_count == MAX_SHORTCUTS) {
+        log_appendf(5, " %s/%s: Trying to bind too many shortcuts. Max is %i.", bind->section_info->name, bind->name, MAX_SHORTCUTS);
+        printf("%s/%s: Trying to bind too many shortcuts. Max is %i.\n", bind->section_info->name, bind->name, MAX_SHORTCUTS);
+        fflush(stdout);
+        has_init_problem = 1;
+        return;
+    }
+
+    if(keycode == SDLK_UNKNOWN && scancode == SDL_SCANCODE_UNKNOWN && (!character || !character[0])) {
+        printf("Attempting to bind shortcut with no key. Skipping.\n");
+        return;
+    }
+
+    int i = bind->shortcuts_count;
+    bind->shortcuts[i].scancode = scancode;
+    bind->shortcuts[i].modifier = modifier;
+    bind->shortcuts[i].character = character;
+    bind->shortcuts_count++;
+}
+
+static void init_section(keybind_section_info* section_info, const char* name, const char* title, enum page_numbers page)
+{
+    section_info->is_active = 0;
+    section_info->name = name;
+    section_info->title = title;
+    section_info->page = page;
+    section_info->page_matcher = NULL;
+}
+
+static void set_shortcut_text(keybind_bind* bind)
+{
+    char* out[MAX_SHORTCUTS];
+    int is_first_shortcut = 1;
+
+    for(int i = 0; i < MAX_SHORTCUTS; i++) {
+        out[i] = NULL;
+        if(i >= bind->shortcuts_count) continue;
+
+        keybind_shortcut* sc = &bind->shortcuts[i];
+
+        const char* ctrl_text = (sc->modifier & KMOD_CTRL) ? "Ctrl-" : "";
+        const char* alt_text = (sc->modifier & KMOD_LALT) ? "Alt-" : "";
+        const char* shift_text = (sc->modifier & KMOD_SHIFT) ? "Shift-" : "";
+
+        char* key_text = NULL;
+
+        if(sc->character[0]) {
+            key_text = strdup(sc->character);
+        } else if(sc->scancode != SDL_SCANCODE_UNKNOWN) {
+            switch(sc->scancode) {
+                case SDL_SCANCODE_RETURN:
+                    key_text = strdup("Enter");
```

Comment:
There's no scancode for "enter". [scancodes](https://wiki.libsdl.org/SDL2/SDL_Scancode)

Reply:
Yep because it's usually only present on "extended" 101-key ANSI keyboards on the number pad. (some keyboards even mislabel the return key as an enter key as well, which is just plain wrong)

It is important to note that there really are keyboards with separate physical enter and return keys in the main section, notably old Apple portables. see the the Fn-Return issue for an example.

Reply:
Hmm okay so what should I do? I added this code because I wanted to help text to match how it looked before custom keybinds. There it says "Enter" and not "Return".

Reply:
Just change the help text to be correct then :)